flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/ContinuousHistogram.scala

```diff
@@ -0,0 +1,337 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.math
+
+import scala.Double.MaxValue
+import scala.collection.mutable
+
+/** Implementation of a continuous valued online histogram
+  * Adapted from Ben-Haim and Yom-Tov's Streaming Decision Tree Algorithm
+  * Refer http://www.jmlr.org/papers/volume11/ben-haim10a/ben-haim10a.pdf
+  *
+  */
+case class ContinuousHistogram(
+    capacity: Int,
+    min: Double,
+    max: Double)
+  extends OnlineHistogram {
+
+  require(capacity > 0, "Capacity should be a positive integer")
+  require(lower < upper, "Lower must be less than upper")
+
+  val data = new mutable.ArrayBuffer[(Double, Int)]()
+
+  /** Adds a new instance
+    *
+    * @param p value to be added
+    */
+  override def add(p: Double): Unit = {
+    require(p > lower && p < upper, p + " not in (" + lower + "," + upper + ")")
+    // search for the index where the value is just higher than p
+    val search = find(p)
+    // add the new value there, shifting everything to the right
+    data.insert(search, (p, 1))
+    // If we're over capacity or any two elements are within 1e-9 of each other, merge.
+    // This will take care of the case if p was actually equal to some value in the histogram and
+    // just increment the value there
+    mergeElements()
+  }
+
+  /** Merges the histogram with h and returns a histogram with capacity B
+    *
+    * @param h histogram to be merged
+    * @param B capacity of the resultant histogram
+    * @return Merged histogram with capacity B
+    */
+  override def merge(h: OnlineHistogram, B: Int): ContinuousHistogram = {
+    h match {
+      case temp: ContinuousHistogram => {
+        val m: Int = bins
+        val n: Int = temp.bins
+        var i, j: Int = 0
+        val mergeList = new mutable.ArrayBuffer[(Double, Int)]()
+        while (i < m || j < n) {
+          if (i >= m) {
+            mergeList.+=((temp.getValue(j), temp.getCounter(j)))
```

Comment:
Can change to `mergeList += ((temp.getValue(j), temp.getCounter(j)))` to improve readability.